sha512/src/main/kotlin/org/komputing/khash/sha512/Sha512.kt

```diff
@@ -199,65 +129,45 @@ object Sha512 {
     private fun padMessage(input: ByteArray): ByteArray {
         // Need to append at least 17 bytes (16 for length of the message, and 1 for the '1' bit)
         // then fill with 0s until multiple of 128 bytes
-        var size = input.size + 17
-
-        val remainder = size % 128
-        if (remainder != 0) {
-            size += 128 - remainder
+        val size = (input.size + 17).let {
+            if (it % 128 == 0) it else it + 128 - it % 128
         }
+
+        val len = input.size * 8L
         // The padded byte array
-        val out = ByteArray(size)
-        // Copy over the old stuff
-        for (i in input.indices) {
-            out[i] = input[i]
+        return ByteArray(size) { i ->
+            when {
+                i < input.size -> input[i]
+                i == input.size -> 0x80.toByte()
+                i == size - 1 -> len.toByte()
+                i == size - 2 -> (len ushr 8).toByte()
+                i == size - 3 -> (len ushr 16).toByte()
+                i == size - 4 -> (len ushr 24).toByte()
+                else -> 0
+            }
         }
-        // Add the '1' bit
-        out[input.size] = 0x80.toByte()
-        // And put original length of input at the end of our padded input
-        val len = input.size * 8L
-        out[size - 1] = len.toByte()
-        out[size - 2] = (len ushr 8).toByte()
-        out[size - 3] = (len ushr 16).toByte()
-        out[size - 4] = (len ushr 24).toByte()
-        return out
+
     }
 
     // Converts the byte array input starting at index j into a long
-    private fun arrToLong(input: ByteArray, j: Int): Long {
-        var v: Long = 0
-        for (i in 0..7) {
-            v = (v shl 8) + input[i + j].toUByte().toLong()
-        }
-        return v
+    private fun arrToLong(input: ByteArray, j: Int) = (0..7).fold(0L) { acc, i ->
+        (acc shl 8) + input[i + j].toUByte().toLong()
     }
 
     // Converts the byte array input into blocks of longs
-    private fun toBlocks(input: ByteArray): Array<LongArray> { // a block has: 1024 bits = 128 bytes = 16 longs
-        val blocks = Array(input.size / 128) {
-            val block = LongArray(16)
-            for (j in 0..15) { // Set the block value to the correct one
-                block[j] = arrToLong(input, it * 128 + j * 8)
-            }
-            block
-        }
-        return blocks
+    // a block has: 1024 bits = 128 bytes = 16 longs
+    private fun toBlocks(input: ByteArray) = Array(input.size / 128) {
+        LongArray(16) { j -> arrToLong(input, it * 128 + j * 8) }
     }
 
     // Calculates the expanded message blocks W0-W79
-    private fun calculateMessageBlocks(M: Array<LongArray>): Array<LongArray> {
-        val W = Array(M.size) { LongArray(80) }
-        // For each block in the input
+    private fun calculateMessageBlocks(M: Array<LongArray>) = Array(M.size) { i ->
+        LongArray(80) { j -> if (j < 16) M[i][j] else 0 }
+    }.apply {
```

Comment:
What is the motivation to do it this way?

Reply:
It's more concise and IMHO better expressign what it is doing